### PR TITLE
[Core][V1] Support trace_decode_token_ids for deterministic decode replay

### PR DESCRIPTION
**Pull request by zllion**
Add inference trace-replay: a new SamplingParams.trace_decode_token_ids field that forces the v1 sampler to emit a predetermined sequence of decode token IDs, step by step, instead of sampling. Real logprobs and token ranks are still computed from the unmodified logit distribution.

The sampler injects each trace token after sampling (so logprobs stay real); InputBatch tracks the trace per request in parallel with spec_token_ids; and EngineCore.add_request caps generation to the trace length / forces ignore_eos, falling back to normal sampling (with a warning) for incompatible modes (speculative decoding, n > 1).

Tests cover the SamplingParams field, sampler injection (including async -1 placeholders and past-end-of-trace), and engine-core admission. A demo example script is added under examples/generate/.

## Purpose

Add **inference trace-replay**: a new `SamplingParams.trace_decode_token_ids`
field that forces the v1 sampler to emit a predetermined sequence of decode
token IDs, step by step, instead of sampling. Real logprobs and token ranks
are still computed from the unmodified logit distribution at each step.

**Motivation.** The goal is to make a request follow a *fixed* token trace and
return the logprobs the model assigns to that exact trace. This is useful for:

- **Quantifying differences between settings/backends.** Replaying the same
  trace under different configs (dtype, quantization, attention/MoE backend,
  TP size, etc.) and comparing per-token logprobs gives a direct, apples-to-apples
  measure of numerical divergence on identical token sequences.
- **RL training.** It lets you recompute logprobs for a rollout's exact tokens
  under the training engine, so the rollout↔train logprob gap can be measured
  and minimized — without re-sampling and drifting onto a different trajectory.

**Why decode-replay and not prefill-only scoring?** Computing logprobs by
prefilling the full (prompt + response) sequence does *not* faithfully
reproduce the decode-time distribution: many models run **different operators /
kernels in the prefill path vs. the decode path** (e.g. attention and MoE
kernels differ between the two phases). To measure the logprobs a model would
actually produce while *decoding*, the tokens must go through the decode path
one step at a time — which is exactly what trace-replay does.

### How it works

1. Set `SamplingParams.trace_decode_token_ids` to the list of decode token IDs
   to force.
2. The sampler injects each trace token **after** sampling
   (`Sampler._inject_trace_tokens` in `vllm/v1/sample/sampler.py`), so logprobs
   stay real — they are computed from the original logits, before injection.
3. `InputBatch` tracks the trace per request in parallel with `spec_token_ids`
   (`vllm/v1/worker/gpu_input_batch.py`); the per-request step index counts real
   output tokens, skipping the trailing `-1` async-scheduling placeholder.
4. `EngineCore.add_request` caps generation to the trace length and forces
   `ignore_eos`, so EOS tokens inside the trace do not halt generation early. It
   falls back to normal sampling (with a warning) for incompatible modes
   (speculative decoding, `n > 1`).

### Not a duplicate

No open PR addresses this — searched the repo for open PRs matching
`trace_decode_token_ids` and `trace`/`replay` in title (0 results). The closest
existing features are `logprob_token_ids` (only *scores* a fixed token set,
doesn't force the output sequence) and `spec_token_ids` (speculative decoding,
different purpose); neither forces a predetermined decode trace.

## Test Plan

```bash
# Unit tests (CPU-only, no GPU required)
.venv/bin/python -m pytest \
    tests/v1/sample/test_trace_replay.py \
    tests/v1/engine/test_engine_core_trace_replay.py -v

# Lint / type-check as in CI
pre-commit run --all-files
pre-commit run mypy-3.12 --all-files --hook-stage manual
```

- `tests/v1/sample/test_trace_replay.py` — `SamplingParams` field defaults,
  clone preservation, and validation (empty / negative / non-int rejected);
  sampler injection covering step 0, mixed per-request steps, past-end-of-trace
  (no-op), the async `-1` placeholder, and the unset/none path.
- `tests/v1/engine/test_engine_core_trace_replay.py` — engine-core admission:
  generation bounds capped to trace length + `ignore_eos` forced; fallback to
  normal sampling for speculative decoding and `n > 1`.
- Demo: `examples/generate/trace_replay_offline.py` — greedy-generate to capture
  decode tokens, then replay them and assert a token-by-token match while
  printing per-token logprobs.

## Test Result

```text
$ .venv/bin/python -m pytest tests/v1/sample/test_trace_replay.py \
      tests/v1/engine/test_engine_core_trace_replay.py
tests/v1/sample/test_trace_replay.py ..............                      [ 82%]
tests/v1/engine/test_engine_core_trace_replay.py ...                     [100%]
======================= 17 passed, 16 warnings in 0.14s ========================
```

`pre-commit run --all-files` and `mypy-3.12` (CI hook stage) pass on the changed
files (ruff-check, ruff-format, SPDX headers, forbidden-imports, etc.).

> ⚠️ Unit tests were run on a CPU-only box; the end-to-end demo
> (`examples/generate/trace_replay_offline.py`) should be run on a GPU to confirm
> token-by-token replay before merge.

> **AI assistance disclosure:** developed with AI assistance (Claude). The
> submitter has reviewed every changed line and is responsible for the change
> end-to-end.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/v1/sample/sampler.py [hunk 1]
```diff
@@ -355,6 +362,40 @@ def gather_logprobs(
...
+        for req_idx, trace_tokens in enumerate(trace_decode_token_ids):
+            if not trace_tokens:
+                continue
+            produced = output_token_ids[req_idx]
+            # At most one trailing -1 placeholder (no spec decode in trace mode).
+            step_idx = len(produced)
+            if step_idx and produced[-1] == -1:
+                step_idx -= 1
+            if step_idx < len(trace_tokens):
+                sampled[req_idx] = trace_tokens[step_idx]
```
[the remark this answers is not in the document]
Comment (zllion) [reply vllm/v1/sample/sampler.py]: It's already checked in sampling_params.py

**Comment by aoshen02** [comment]
We will shift to model runner v2, could you also implement it in model runner v2?

File: vllm/v1/engine/core.py [hunk 2]
```diff
@@ -369,6 +369,40 @@ def get_kv_cache_group_metadata(self) -> list[dict[str, int | str | None]]:
...
+                "trace-replay is incompatible with %s; "
+                "falling back to normal sampling for request %s.",
+                conflict_reason,
+                request.request_id,
+            )
+            request.sampling_params = msgspec.structs.replace(
+                params, trace_decode_token_ids=None
+            )
+        else:
+            new_max = len(params.trace_decode_token_ids)
```
Comment (aoshen02) [inline vllm/v1/engine/core.py]: Could we clear the cached EOS state? I think ignore_eos is not used when check stop?

File: PR_DESCRIPTION.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+field that forces the v1 sampler to emit a predetermined sequence of decode
+token IDs, step by step, instead of sampling. Real logprobs and token ranks
+are still computed from the unmodified logit distribution at each step.
+
+**Motivation.** The goal is to make a request follow a *fixed* token trace and
+return the logprobs the model assigns to that exact trace. This is useful for:
+
+- **Quantifying differences between settings/backends.** Replaying the same
+  trace under different configs (dtype, quantization, attention/MoE backend,
+  TP size, etc.) and comparing per-token logprobs gives a direct, apples-to-apples
```
Comment (aoshen02) [inline PR_DESCRIPTION.md]: why pr description here?

**Comment by aoshen02** [comment]
@codex review

**Comment by aoshen02** [comment]
@codex review

File: vllm/v1/worker/gpu/sample/trace_replay.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        if trace is not None:
+            self.trace_len.np[req_idx] = len(trace)
+            self.trace_token_ids.stage_write(req_idx, 0, trace)
+            self.use_trace[req_idx] = True
+        else:
+            self.trace_len.np[req_idx] = 0
+            self.use_trace[req_idx] = False
+
+    def apply_staged_writes(self) -> None:
+        self.trace_len.copy_to_uva()
```
Comment (aoshen02) [inline vllm/v1/worker/gpu/sample/trace_replay.py:43]: can we eliminate the cost here.

File: vllm/v1/worker/gpu/sample/sampler.py [hunk 5]
```diff
@@ -53,6 +54,7 @@ def __init__(
         self.bad_words_state = BadWordsState(req_states)
         self.logprob_token_ids_state = LogprobTokenIdsState(max_num_reqs, device)
         self.thinking_budget_state = ThinkingBudgetState(req_states, reasoning_config)
+        self.trace_replay_state = TraceReplayState(req_states)
```
Comment (njhill) [inline vllm/v1/worker/gpu/sample/sampler.py]: I think we should have a config parameter controlling whether or not this is created since it's such a niche feature and we are allocating a large tensor.

File: vllm/v1/worker/gpu/sample/sampler.py [hunk 6]
```diff
@@ -111,6 +115,16 @@ def __call__(
...
         )
 
+        # Overwrite sampled tokens with the replay trace (if any) before
+        # computing logprobs, so logprobs and ranks reflect the real
+        # distribution of the forced token.
+        self.trace_replay_state.apply_trace(
+            sampled,
+            idx_mapping,
+            self.req_states.total_len.gpu,
+            self.req_states.prompt_len.gpu,
```
Comment (njhill) [inline vllm/v1/worker/gpu/sample/sampler.py]: `TraceReplayState` can also store a ref to these two so they don't need to be passed every time?

File: vllm/sampling_params.py [hunk 7]
```diff
@@ -272,6 +272,10 @@ class SamplingParams(
     """Token IDs that stop the generation when they are generated. The returned
     output will contain the stop tokens unless the stop tokens are special
     tokens."""
+    trace_decode_token_ids: list[int] | None = None
+    """If provided, forces the engine to emit this predetermined sequence of
+    token IDs during decoding instead of sampling randomly. Real logprobs are
+    still computed. Conflict checking is performed at the engine level."""
```
Comment (njhill) [inline vllm/sampling_params.py]: I think we should reorg the order of params so that debugging/RL-specific ones like this are grouped separately. It's not one that would be used for prod runtime.

File: vllm/v1/engine/input_processor.py [hunk 8]
```diff
@@ -340,6 +356,8 @@ def process_inputs(
             )
             if self.tokenizer is not None:
                 sampling_params.update_from_tokenizer(self.tokenizer)
+            if sampling_params.trace_decode_token_ids:
```
Comment (njhill) [inline vllm/v1/engine/input_processor.py:375]: can move this check inside the method

File: vllm/v1/worker/gpu/sample/trace_replay.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+        )
+        self.trace_len = UvaBackedTensor(max_num_reqs, dtype=torch.int32)
+        # CPU mirror used to skip the kernel launch when no request replays.
+        self.use_trace = np.zeros(max_num_reqs, dtype=bool)
+
+    def add_request(self, req_idx: int, sampling_params: SamplingParams) -> None:
+        trace = sampling_params.trace_decode_token_ids
+        if trace is not None:
+            self.trace_len.np[req_idx] = len(trace)
+            self.trace_token_ids.stage_write(req_idx, 0, trace)
```
[the remark this answers is not in the document]
Comment (njhill) [reply vllm/v1/worker/gpu/sample/trace_replay.py:38]: This looks like legit comment to address

File: vllm/v1/engine/input_processor.py [hunk 10]
```diff
@@ -156,6 +156,22 @@ def _validate_params(
...
 
+    @staticmethod
+    def _normalize_trace_replay_params(sampling_params: SamplingParams) -> None:
+        """Apply trace replay's generation semantics to request-local params."""
+        trace_token_ids = sampling_params.trace_decode_token_ids
+        assert trace_token_ids
+
+        # Apply this after the generation config so its EOS token cannot stop
+        # replay before the trace is exhausted.
+        sampling_params.max_tokens = len(trace_token_ids)
```
Comment (njhill) [inline vllm/v1/engine/input_processor.py]: This should be clamped to max_model_len

File: examples/generate/trace_replay_offline.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+How it works:
+  1. Set ``SamplingParams.trace_decode_token_ids`` to the list of decode
+     token IDs you want to force.
+  2. The engine will output exactly those tokens and stop; ``max_tokens``
+     is ignored.  EOS tokens inside the trace sequence do **not** halt
+     generation early.
+
+Incompatible features (fall back to normal sampling with a warning):
+  * Speculative decoding
+  * n > 1
```
Comment (njhill) [inline examples/generate/trace_replay_offline.py]: This sounds incorrect. Requests should fail if this is specified and unsupported rather than silently falling back.

**Review by njhill — COMMENTED** [review]
In addition to the inline comments, it would be preferred if we only add this to MRV2, since we will be deprecating MRV1 very soon.

File: vllm/v1/worker/gpu/sample/sampler.py (hunk 5 above)
Comment (zllion) [reply vllm/v1/worker/gpu/sample/sampler.py]: Done

File: vllm/v1/engine/input_processor.py (hunk 8 above)
Comment (zllion) [reply vllm/v1/engine/input_processor.py:375]: Done

File: vllm/sampling_params.py (hunk 7 above)
Comment (zllion) [reply vllm/sampling_params.py]: Done

**Comment by aoshen02** [comment]
@njhill could you take a second look?

**Comment by zllion** [comment]
Sorry for the force push — I'd accidentally pushed a couple of commits before self-reviewing them. 
I rewrote the branch so that each review comment maps to exactly one commit, which should make it easier to trace what addressed what. The commit that drops the MRV1 implementation is deliberately last, since it's by far the largest change. Nothing was dropped in the rewrite; the only side effect is that some inline threads above are now marked outdated.

File: vllm/v1/worker/gpu/sample/sampler.py [hunk 12]
```diff
@@ -128,6 +137,12 @@ def __call__(
             return_logprobs=return_logprobs,
         )
 
+        # Overwrite sampled tokens with the replay trace (if any) before
+        # computing logprobs, so logprobs and ranks reflect the real
+        # distribution of the forced token.
+        if self.trace_replay_state is not None:
+            self.trace_replay_state.apply_trace(sampled, idx_mapping)
```
Comment (njhill) [inline vllm/v1/worker/gpu/sample/sampler.py]: Trim a bit
```suggestion
        if self.trace_replay_state is not None:
            # Overwrite sampled tokens with the replay trace up-front so that
            # computed logprobs reflect the real distribution of the forced token.
            self.trace_replay_state.apply_trace(sampled, idx_mapping)
```

File: vllm/v1/engine/input_processor.py [hunk 13]
```diff
@@ -128,6 +128,18 @@ def _validate_params(
...
+                    raise VLLMValidationError(
+                        "trace_decode_token_ids is set but trace replay is not "
+                        "enabled. Start the engine with --enable-trace-replay "
+                        "to use it."
+                    )
+                if not self.use_v2_model_runner:
+                    raise VLLMValidationError(
+                        "trace_decode_token_ids is only supported by model "
+                        "runner V2, but this engine is running model runner V1."
+                    )
```
Comment (njhill) [inline vllm/v1/engine/input_processor.py]: Could you move this check to config validation (in `config/vllm.py`), then it's not needed here.

File: vllm/sampling_params.py [hunk 14]
```diff
@@ -373,6 +373,12 @@ class SamplingParams(
     '\\emoji \\emoji \\emoji ...'). This feature can detect such behavior
     and terminate early, saving time and tokens."""
 
+    # Debugging / RL-specific parameters. Not intended for production serving.
+    trace_decode_token_ids: list[int] | None = None
+    """If provided, forces the engine to emit this predetermined sequence of
+    token IDs during decoding instead of sampling randomly. Real logprobs are
+    still computed. Conflict checking is performed at the engine level."""
```
Comment (njhill) [inline vllm/sampling_params.py:379]: Could we move `routed_experts_prompt_start` next to this (again so that the RL params are separated/grouped).

**Review by njhill — COMMENTED** [review]
Thanks @zllion @aoshen02 ... just a couple more small comments

**Review by njhill — APPROVED** [review]
Thanks @aoshen02 @zllion

File: vllm/v1/engine/input_processor.py (hunk 13 above)
Comment (zllion) [reply vllm/v1/engine/input_processor.py]: Done

File: vllm/sampling_params.py (hunk 14 above)
Comment (zllion) [reply vllm/sampling_params.py:379]: Done
